src/main/java/com/ning/http/client/RequestBuilderBase.java

```diff
@@ -373,23 +373,28 @@ public T addCookie(Cookie cookie) {
         return derived.cast(this);
     }
 
-    public T addOrReplaceCookie(Cookie cookie) {
-        String cookieKey = cookie.getName();
-        boolean replace = false;
-        int index = 0;
+    public void cleanExpiredCookies() {
         lazyInitCookies();
+
         for (Cookie c : request.cookies) {
-            if (c.getName().equals(cookieKey)) {
-                replace = true;
-                break;
+            if (c.hasExpired()) {
+                request.cookies.remove(c);
             }
+        }
+    }
 
-            index++;
+    public T addOrReplaceCookie(Cookie cookie) {
+        int index = -1;
+        lazyInitCookies();
+
+        cleanExpiredCookies();
+        if (request.cookies.contains(cookie)) {
+            index = request.cookies.indexOf(cookie);
         }
-        if (replace)
-            request.cookies.set(index, cookie);
-        else
+        // if cookie not found, then add it
```

Comment:
shouldn't this be in an else as the -1 is changed only if request.cookies.contains(cookie) is false?